nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Connection.java

```diff
@@ -291,6 +346,28 @@ private void readFrame() {
         if (frameHeader.length() == 0) {
             frameInProgress = BufferData.empty();
         } else {
+            // Flow-control: reading frameHeader.length() bytes from HTTP2 socket for known stream ID.
+            int length = frameHeader.length();
+            if (length > 0) {
+                int streamId = frameHeader.streamId();
+                if (streamId > 0 && frameHeader.type() != Http2FrameType.HEADERS) {
+                    // Stream ID > 0: update conenction and stream
+                    FlowControl.Inbound inboundFlowControl = stream(streamId)
+                            .stream()
+                            .inboundFlowControl();
+                    inboundFlowControl.decrementWindowSize(length);
+                    LOGGER.log(System.Logger.Level.INFO, () -> String.format(
+                            "SRV IFC: Full decrement %d-> %d", length,
+                            inboundFlowControl.getRemainingWindowSize()));
+                } else {
+                    // Stream ID == 0: update connection only
+                    inboundWindowSize.decrementWindowSize(length);
+                    LOGGER.log(System.Logger.Level.INFO, () -> String.format(
+                            "SRV IFC: Conn decrement %d -> %d",
+                            length, inboundWindowSize.getRemainingWindowSize()));
+                }
+            }
```

Comment:
This doesn't belong here, decrement needs to be in `io.helidon.nima.http2.webserver.Http2Stream#readEntityFromPipeline`